Convert bpf endpoint config macros to load time config

fristonio opened the pull request:
Convert BPF macros to load time config:
* `HOST_EP_ID`
* `DIRECT_ROUTING_DEV_IFINDEX`

Parent Issue: https://github.com/cilium/cilium/issues/38370

fristonio committed:
bpf: convert HOST_EP_ID macro to load time config

Signed-off-by: Deepesh Pathak <deepeshpathak09@gmail.com>

fristonio committed:
bpf: convert DIRECT_ROUTING_DEV_IFINDEX macro to load time config

Signed-off-by: Deepesh Pathak <deepeshpathak09@gmail.com>

ti-mo commented on bpf/include/bpf/config/node.h:
```diff
@@ -22,3 +22,7 @@ NODE_CONFIG(__u32, trace_payload_len, "Length of payload to capture when tracing
+
+NODE_CONFIG(__u32, direct_routing_dev_ifindex, "Index of the interface used to connect nodes in the cluster.")
+#define DIRECT_ROUTING_DEV_IFINDEX CONFIG(direct_routing_dev_ifindex) /* Backwards compatibility */
+ASSIGN_CONFIG(__u32, direct_routing_dev_ifindex, 0)
```
These are plain variables and zero-initialized, so this isn't needed.

ti-mo commented on bpf/include/bpf/config/endpoint.h:
```diff
@@ -18,3 +18,6 @@ DECLARE_CONFIG(__u32, security_label, "The endpoint's security label")
 #define SECLABEL_IPV6 SECLABEL
+
+DECLARE_CONFIG(__u16, host_ep_id, "The host endpoint's security ID")
+#define HOST_EP_ID CONFIG(host_ep_id) /* Backwards compatibility */
```
(on mobile) Are there a lot of references to HOST_EP_ID? Up to a dozen, it's not too invasive to rewrite them.

fristonio replied:
Actually yeah, not a lot of references. I have cleaned up references to this macro, but there is a preprocessor that check the definition of this macro to decide effective Endpoint Id for policy maps. I have refactored that evaluation to get rid of this macro. The change is in a separate commit, let me know if it makes sense.

ti-mo commented on bpf/include/bpf/config/endpoint.h:
```diff
@@ -18,3 +18,6 @@ DECLARE_CONFIG(__u32, security_label, "The endpoint's security label")
 #define SECLABEL_IPV4 SECLABEL
 #define SECLABEL_IPV6 SECLABEL
+
+DECLARE_CONFIG(__u16, host_ep_id, "The host endpoint's security ID")
```
Could this go directly into bpf_lxc? Or does the host ep itself also need to know its own id?

fristonio replied:
Both bpf_host and bpf_lxc depend on host endpoint id so I guess `endpoint.h` is the right place? For `bpf_host` there are two usage:
1. as event source - https://github.com/cilium/cilium/blob/8cc584c71983bdef961873f59a08a6c510272f82/bpf/bpf_host.c#L14
2. As endpoint_id in policy stats map
https://github.com/cilium/cilium/blob/8cc584c71983bdef961873f59a08a6c510272f82/bpf/lib/policy.h#L30

ti-mo replied:
Yep, sounds good. :+1:

ti-mo commented on pkg/datapath/loader/config.go:
```diff
@@ -23,5 +23,9 @@ func nodeConfig(lnc *datapath.LocalNodeConfiguration) config.Node {
 	node.TracePayloadLen = uint32(option.Config.TracePayloadlen)
 	node.TracePayloadLenOverlay = uint32(option.Config.TracePayloadlenOverlay)
 
+	if option.Config.DirectRoutingDeviceRequired(lnc.KPRConfig) && lnc.DirectRoutingDevice != nil {
```
Would it make sense to only gate this on the presence of the direct routing device? I assume there's some upstream logic that only populates it when direct routing is enabled, and we should keep this config glue as light as possible.

fristonio replied:
Doesn't look like it. The config is populated from device table irrespective of whether DirectRoutingDevice is required or not. But it seems fairly straightforward to handle this upstream. I have the corresponding changes in the latest commit, but not sure if there can be side effects. Let me know what you think.

fristonio replied:
@dylandreimerink can you take a look at this commit? Just to make sure there are no unintended side effects? https://github.com/cilium/cilium/pull/40430/commits/ce709caaa9b3210c78dd7b82afb95fd646aa3a64

ti-mo reviewed (commented):
Thanks! Left a few considerations.

fristonio committed:
bpf: remove HOST_EP_ID macro and use scoped macros for each program

Signed-off-by: Deepesh Pathak <deepeshpathak09@gmail.com>

fristonio committed:
datapath: move DirectRoutingDevice validation to orchestrator

This commit refactors the evaluation of DirectRoutingDevice requirement
to orchestrator simplifying the usage during load time.
When reconciling LocalNodeConfig error out if Direct routing device is
required but not yet configured.

Signed-off-by: Deepesh Pathak <deepeshpathak09@gmail.com>

dylandreimerink reviewed (approved):
Taking over review for cilium/loader since Timo is not available for a bit.

Changes look good to me. I believe all of Timo's feedback has been processed. Thank you!

derailed reviewed (approved):
@fristonio Nice work!

ldelossa reviewed (approved):
LGTM
